cocos/2d/renderer/mesh-buffer.ts

```diff
@@ -104,7 +104,10 @@ export class MeshBuffer {
         this._attributes = attrs;
         this._iaInfo = new InputAssemblerInfo(this.attributes, this.vertexBuffers, this.indexBuffer);
 
-        this._reallocBuffer();
+        // for recycle pool using purpose --
+        if (!this.vData || !this.iData)
```

Comment:
I understand repetitive allocation could be improved.
But if we only remove allocation, with `_initVDataCount` reset, it could cause mismatch and due to unnecessary reallocation. Please consider reset `_initVDataCount` accordingly.

```
let count = this._initVDataCount;
this._initVDataCount = Math.floor(count / this._vertexFormatBytes) * this._vertexFormatBytes;
```

Reply:
More seriously, if the new `_vertexFormatBytes` is bigger than the previous one, and we re-initialize it, the current implementation could cause out of bound access.

Reply:
_initIDataCount and _initVDataCount is not a value to determine the length of the buffer you created. It is just a variable for you to calculating a new valid buffer size incase you request the buffer with bigger length. Please check the code in 
the method `public request (vertexCount = 4, indicesCount = 6)`

Reply:
Note the we reuse this.vData and this.iData, the buffers length will not be change until you request a bigger length. Reset() only need to reset 
this.byteOffset = 0;
this.indicesOffset = 0;
so we can start to write to the buffers from the first byte when we reuse it from pool.

Reply:
Ok, indeed we are using byteOffset and actual ArrayBuffer byteLength to check availability, so won't cause any out of bound issue. But the `_initVDataCount` value should be matching vData byteLength, that's just an implication of their relationship.

Imaging the following case
1. The buffer was initialized to contain 256 vertices
2. The buffer have been requested multiple times, till it grow to 1024 vertices length
3. The recycle pool reseted
4. Then the same buffer is retrieved and  re-initialized
5. The `_initVDataCount` reset to 256 while the buffer still have 1024 vertices long
6. While requested a buffer bigger than 1024 vertices, the while loop to calculate correct `_initVDataCount` will be running 2 more times than needed.

Reply:
in the case above `_initVDataCount` will be double in while loop from a beginning value (`256 * this._vertexFormatBytes`) until the byteLength (`this._initVDataCount * 4`) bigger or equal the new required byteLength (`this.byteOffset + vertexCount * this._vertexFormatBytes`). Not sure it is the best way to calculate the new require byte length but it is from the original code already.